# Análisis general
A continuación se realizará un análisis general de las relaciones que existen entre las diferentes columnas del dataset.

In [27]:
import pandas as pd

In [28]:
# Replico la estructura de los tweets.
tweets = spotify_songs = pd.read_csv('Data/train.csv')
tweets['length'] = tweets['text'].str.len()
tweets['located'] = tweets['location'].notnull().astype(int)
tweets['has_keyword'] = tweets['keyword'].notnull().astype(int)
tweets['keyword_length'] = tweets['keyword'].str.len()
tweets['keyword_length'].fillna(0, inplace=True)
tweets['tweet_length'] = tweets['length'] + tweets['keyword_length']
tweets.tail()

id keyword location  \
7608  10869     NaN      NaN   
7609  10870     NaN      NaN   
7610  10871     NaN      NaN   
7611  10872     NaN      NaN   
7612  10873     NaN      NaN   

                                                   text  target  length  \
7608  Two giant cranes holding a bridge collapse int...       1      83   
7609  @aria_ahrary @TheTawniest The out of control w...       1     125   
7610  M1.94 [01:04 UTC]?5km S of Volcano Hawaii. htt...       1      65   
7611  Police investigating after an e-bike collided ...       1     137   
7612  The Latest: More Homes Razed by Northern Calif...       1      94   

      located  has_keyword  keyword_length  tweet_length  
7608        0            0             0.0          83.0  
7609        0            0             0.0         125.0  
7610        0            0             0.0          65.0  
7611        0            0             0.0         137.0  
7612        0            0             0.0          94.0

## Describe
Se realiza un describe para realizar un rápido análisis de las nuevas columnas numéricas que se agregaron: cantidad, media, desvío estándar, mínimo, 25%, 50%, 75% y máximo.

In [56]:
tweets.describe()

id      target       length      located  has_keyword  \
count   7613.000000  7613.00000  7613.000000  7613.000000  7613.000000   
mean    5441.934848     0.42966   101.037436     0.667280     0.991987   
std     3137.116090     0.49506    33.781325     0.471218     0.089160   
min        1.000000     0.00000     7.000000     0.000000     0.000000   
25%     2734.000000     0.00000    78.000000     0.000000     1.000000   
50%     5408.000000     0.00000   107.000000     1.000000     1.000000   
75%     8146.000000     1.00000   133.000000     1.000000     1.000000   
max    10873.000000     1.00000   157.000000     1.000000     1.000000   

       keyword_length  tweet_length  
count     7613.000000   7613.000000  
mean         8.667148    109.704584  
std          3.534058     34.492177  
min          0.000000      8.000000  
25%          6.000000     86.000000  
50%          8.000000    116.000000  
75%         10.000000    142.000000  
max         21.000000    163.000000

### id
Como sabemos id es un identificador único para cada tweet, no tiene sentido alguno analizar sus valores numéricos.

### Target
Es la única variable numérica que se analizó en la primera sección "Estructura del set de datos". De esa sección se sabe que los tweets que no refieren a un siniestro real (0) son mayores a los que si refieren a un siniestro real (1), aquí queda en evidencia en la media de la columna, ya que es más cercana a 0 que a 1.

### length
La longitud de caracteres máxima de los tweets es de 280 caracteres.\
La longitud media de los tweets del set de datos es de solo 101 caracteres mientras que su máximo es de 157 caracteres, valor que está realmente alejado de los 280 caracteres de longitud máxima permitida por Twitter pero, a su vez, es mayor que los 140 caracteres que Twitter permitía en sus primeras versiones.\
El tweet con menor longitud es de 7 caracteres, una longitud muy pequeña para referirse a algún siniestro. Por esto, analizaremos los tweets cuya longitud es menor o igual a 10:

In [57]:
tweets[tweets['length'] <= 10]

id            keyword         location        text  target  length  \
24      36                NaN              NaN    LOOOOOOL       0       8   
28      40                NaN              NaN   Cooool :)       0       9   
30      44                NaN              NaN    The end!       0       8   
1882  2703            crushed              NaN     Crushed       0       7   
3667  5221           fatality          Nairobi   Fatality!       0       9   
3670  5224           fatality  Rafael castillo    fatality       0       8   
3749  5326               fire              NaN  I See Fire       0      10   
4890  6962           massacre              NaN     Bad day       0       7   
4971  7088           meltdown              NaN    Meltdown       0       8   
5115  7295  nuclear%20reactor              NaN     Err:509       0       7   
6705  9605            thunder              NaN  Thunder???       0      10   

      located  has_keyword  keyword_length  tweet_length  
24          0            0             0.0           8.0  
28          0            0             0.0           9.0  
30          0            0             0.0           8.0  
1882        0            1             7.0          14.0  
3667        1            1             8.0          17.0  
3670        1            1             8.0          16.0  
3749        0            1             4.0          14.0  
4890        0            1             8.0          15.0  
4971        0            1             8.0          16.0  
5115        0            1            17.0          24.0  
6705        0            1             7.0          17.0

La mayoría de los tweets con longitudes pequeñas no son desastres reales a pesar de que varios de ellos tienen textos que se refieren a siniestros reales.

### located
La Columna located indica si el tweet tiene algún tipo de texto que describa su ubicación (1) o no (0).\
Su valor medio nos indica que existen más tweets localizados que aquellos que no tienen localización, ya que el valor medio se encuentra más cercano a 1 que a 0.

### has_keyword
Es una variable numérica pero del tipo booleana (sólo toma valores 0 y 1) que indica si el tweet tiene palabra clave (1) o no (0).\
Su media indica que un gran porcentajes de tweet tienen una palabra clave (99,2%).

### Keyword_length
Indica la longitud de la palabra clave de cada tweet.\
Su valor medio ronda los 8/9 caracteres teniendo en cuenta los tweets que no tienen keyword, entonces el valor medio de la longitud de las palabras claves solo para los tweets que contengan una debe ser superior:

In [58]:
tweets[tweets['has_keyword'] == 1]['keyword_length'].describe()

count    7552.000000
mean        8.737156
std         3.461028
min         4.000000
25%         6.000000
50%         8.000000
75%        10.000000
max        21.000000
Name: keyword_length, dtype: float64

El valor medio de la longitud de las palabras claves aumentó si filtramos aquellos tweets que no poseen palabra clave.\
El valor mínimo es de 0 caracteres que son aquellos tweets sin palabra clave mientras que el máximo es 21 caracteres.\
Se analizan las keywords de longitud mayor o igual a 12 (valor medio + desviación estándar):

In [62]:
tweets[tweets['keyword_length'] > 12]['keyword'].value_counts()

burning%20buildings      37
nuclear%20reactor        36
structural%20failure     35
airplane%20accident      35
buildings%20burning      35
bridge%20collapse        35
suicide%20bomb           35
emergency%20plan         35
natural%20disaster       34
nuclear%20disaster       34
buildings%20on%20fire    33
violent%20storm          33
mass%20murder            33
body%20bagging           33
chemical%20emergency     33
suicide%20bombing        33
emergency%20services     33
forest%20fires           32
mass%20murderer          32
suicide%20bomber         31
first%20responders       29
forest%20fire            19
radiation%20emergency     9
Name: keyword, dtype: int64

Las keywords de más de 12 caracteres de longitud no son palabras sino una oración, por eso se explica su longitud.\
A su vez, el valor mínimo de la longitud de los keyword es de 4 caracteres.\
Se analizan las keywords cuya longitud es menor o igual a 5 (valor medio - desviación estándar):

In [63]:
tweets[tweets['keyword_length'] < 6]['keyword'].value_counts()

harm     41
fear     40
siren    40
sunk     39
fire     38
blaze    38
fatal    38
wreck    37
panic    37
crush    37
ruin     37
death    36
razed    35
storm    35
hail     35
flood    35
blood    35
bomb     34
smoke    34
army     34
riot     34
lava     34
crash    33
drown    32
arson    32
dead     30
Name: keyword, dtype: int64

Las keywords de poca longitud de caracteres describen desastres en su mayoría.

### Tweet_length
Esta columna indica la longitud total de los textos del tweet: longitud del tweet más la longitud de la palabra clave del tweet.\
Su valor medio es la suma de los valores medios de la longitud del tweet más la longitud de la palabra clave del tweet.\
Su valor mínimo es 8 y su valor máximo es 168.\
Se analiza el caso en donde la longitud total es menor o igual a 15:

In [64]:
tweets[tweets['tweet_length'] <= 15]

id   keyword   location            text  target  length  located  \
15       23       NaN        NaN  What's up man?       0      14        0   
16       24       NaN        NaN   I love fruits       0      13        0   
22       33       NaN        NaN     Love skiing       0      11        0   
24       36       NaN        NaN        LOOOOOOL       0       8        0   
28       40       NaN        NaN       Cooool :)       0       9        0   
30       44       NaN        NaN        The end!       0       8        0   
1882   2703   crushed        NaN         Crushed       0       7        0   
3749   5326      fire        NaN      I See Fire       0      10        0   
4735   6734      lava  Venezuela     I LAVA YOU.       0      11        1   
4890   6962  massacre        NaN         Bad day       0       7        0   
7589  10843       NaN        NaN  Omg earthquake       1      14        0   

      has_keyword  keyword_length  tweet_length  
15              0             0.0          14.0  
16              0             0.0          13.0  
22              0             0.0          11.0  
24              0             0.0           8.0  
28              0             0.0           9.0  
30              0             0.0           8.0  
1882            1             7.0          14.0  
3749            1             4.0          14.0  
4735            1             4.0          15.0  
4890            1             8.0          15.0  
7589            0             0.0          14.0

La mayoría de los tweets con longitud total menor o igual a 15 no son tweets sobre desastres reales.